README.md
# ACF: FocusPoint

Adds a new "FocusPoint" field type to [Advanced Custom Fields](https://www.advancedcustomfields.com/) allowing users to select a focal point on images.

If you find this plugin useful, please consider sponsoring the development by clicking the Sponsor button or [directly through PayPal](https://paypal.me/ooksanen).

-----------------------

### Description

Adds a new field type to ACF allowing users to select a focal point on image.

Works similar to a traditional ACF image field, but once an image is selected and a preview is displayed, this plugin allows selecting a focal point by clicking on the image. Focal point is returned in image array as percentage from left/top.

Requires ACF Pro 5

Tested on ACF Pro 5.8.7

-----------------------

### Installation

1. Download and extract the plugin
2. Copy the `acf-focuspoint` folder into your `wp-content/plugins` folder
3. Activate the ACF FocusPoint plugin via the plugins admin page
4. Create a new field via ACF and select the FocusPoint type

-----------------------

### Usage

Usage is very similar to [ACF Image Field](http://www.advancedcustomfields.com/resources/image/). 

1. Select an image

![Screenshot 1](screenshots/screenshot1.png)

2. Select a focal point by clicking the image preview

![Screenshot 2](screenshots/screenshot2.png)

Plugin returns an array with image id and top/left values (percentages) of selected focus point:

```
'image' => array(
	'id'		=> 42,
	'top'  		=> 40.03,
	'left' 		=> 82.79,
)
```
Use these new values as you see fit. For example:

```
<?php 

$image = get_field('image'); 
$image_src = wp_get_attachment_image_src( $image['id'], 'large' );

?>

<style>
#my-image {
	background-image: url('<?php echo $image_src[0]; ?>);
	background-position: <?php echo $image['left'] . '% ' . $image['top']; ?>%;
	background-size: cover;
	height: 300px;
	width: 600px;
}
</style>

<div id="my-image"></div>
```

-----------------------

### Changelog

See [`changelog.md`](https://github.com/ooksanen/acf-focuspoint/blob/master/changelog.md)

-----------------------

### Thanks

Thanks to [Elliot Condon](http://www.elliotcondon.com/) for creating Advanced Custom Fields.


changelog.md
## Changelog

### 1.2.0
* Added options for Library (All / Uploaded to post) as requested in https://github.com/ooksanen/acf-focuspoint/issues/9 and Allowed file types

### 1.1.8
* Bugfix (1.1.7 related)

### 1.1.7
* Fixed a bug where you couldn't select/upload image if previously selected image was deleted

### 1.1.6
* Fixed a bug where multiple FocusPoint images in a repeater saved same top/left values for all images

### 1.1.5
* Fixed saving coordinates when no image selected (https://github.com/ooksanen/acf-focuspoint/issues/8).
* Use minified versions of input.js and input.css

### 1.1.4
* Fixed Required field validation returning false even on non-required field (https://github.com/ooksanen/acf-focuspoint/issues/7).

### 1.1.3
* Fixed Required field validation (https://github.com/ooksanen/acf-focuspoint/issues/6). Note that validation doesn't work in Gutenberg at the moment. See https://github.com/AdvancedCustomFields/acf/issues/113 for more info.

### 1.1.2
* Fixed image select form visibility on initial load and other minor (mostly) CSS issues.

### 1.1.1
* Bug fix: https://github.com/ooksanen/acf-focuspoint/issues/5

### 1.1.0
* Almost complete rewrite of javascript to add proper support for Gutenberg blocks.

### 1.0.4
* Added simple(Classic Editor only?) validation for min/max width, height, and file size

### 1.0.3
* Fixed a bug that prevented deleting focuspoint image
* Started implementing min/max width, height, and file size for image file

### 1.0.2
* Moved focus point selection to new element on top of the image to allow for more fine grained selection even very close to current selection

### 1.0.1
* Remove ACF 4 files / support which was never implemented
* Fixed bug where values weren't saved when using field in ACF Block
* Fixed selector for image hover cursor
* Minor CSS tweaks

### 1.0
* Initial release


acf-focuspoint.php
<?php

/*
Plugin Name: ACF: FocusPoint
Plugin URI: https://github.com/ooksanen/acf-focuspoint/
Description: Adds a new "FocusPoint" field type to Advanced Custom Fields allowing users to select a focal point on images.
Version: 1.2.0
Author: Oskari Oksanen
Author URI: https://oskarioksanen.fi
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html
*/

// exit if accessed directly
if( ! defined( 'ABSPATH' ) ) exit;


// check if class already exists
if( !class_exists('acffp_acf_plugin_focuspoint') ) :

class acffp_acf_plugin_focuspoint {
	
	// vars
	var $settings;
	
	
	/*
	*  __construct
	*
	*  This function will setup the class functionality
	*
	*  @type	function
	*  @date	17/02/2016
	*  @since	1.0.0
	*
	*  @param	void
	*  @return	void
	*/
	
	function __construct() {
		
		// settings
		// - these will be passed into the field class.
		$this->settings = array(
			'version'	=> '1.2.0',
			'url'		=> plugin_dir_url( __FILE__ ),
			'path'		=> plugin_dir_path( __FILE__ )
		);
		
		
		// include field
		add_action('acf/include_field_types', array($this, 'include_field')); // v5
	}
	
	
	/*
	*  include_field
	*
	*  This function will include the field type class
	*
	*  @type	function
	*  @date	17/02/2016
	*  @since	1.0.0
	*
	*  @param	$version (int) major ACF version. Defaults to false
	*  @return	void
	*/
	
	function include_field( $version = false ) {
		
		// load textdomain
		load_plugin_textdomain( 'acffp', false, plugin_basename( dirname( __FILE__ ) ) . '/lang' ); 
		
		// include
		include_once('fields/class-acffp-acf-field-focuspoint-v' . $version . '.php');
	}
	
}


// initialize
new acffp_acf_plugin_focuspoint();


// class_exists check
endif;
	
?>


assets/js/input.js
(function($) {
    acf.fields.focuspoint = acf.field.extend({

        type: 'focuspoint',
        $el: null,

        actions: {
            'ready': 'initialize',
            'append': 'initialize'
        },

        events: {
            'click a[data-name="add"]': 'add',
            'click a[data-name="edit"]': 'edit',
            'click a[data-name="remove"]': 'remove',
            'change input[type="file"]': 'change'
        },

        focus: function() {
            // get elements
            this.$el = this.$field.find('.acf-image-uploader');
            // get options
            this.o = acf.get_data(this.$el);
        },

        initialize: function($el) {

            // add attribute to form
            if (this.o.uploader == 'basic') {
                this.$el.closest('form').attr('enctype', 'multipart/form-data');
            }

            // Cache jquery selectors
            // Values to get/set
            var $id = $el.find('[data-name="acf-focuspoint-img-id"]'),

                // Elements to get/set 
                $fp = $el.find('.acf-focuspoint'),
                $img = $el.find('.acf-focuspoint-img'),
                $selection = $el.find('.focuspoint-selection-layer');

            // Hold/get our values
            var values = {
                id: $id.val(),
                size: $fp.data('preview_size')
            };

            // DOM elements
            var img = $img.get(0);

            $selection.on('click', function(event) {

                var iw = $(this).outerWidth();
                var ih = $(this).outerHeight();
                var px = event.offsetX;
                var py = event.offsetY;
                var y_percentage = Math.round(((py / ih * 100) + Number.EPSILON) * 100) / 100;
                var x_percentage = Math.round(((px / iw * 100) + Number.EPSILON) * 100) / 100;
                $(this).siblings('.focal-point-picker').css({
                    top: y_percentage + '%',
                    left: x_percentage + '%'
                });
                $top = $(this).parents('.acf-focuspoint').children('[data-name="acf-focuspoint-top"]');
                $left = $(this).parents('.acf-focuspoint').children('[data-name="acf-focuspoint-left"]');
                $top.val(y_percentage).trigger('change');
                $left.val(x_percentage).trigger('change');

            });
        },

        add: function() {
            // reference
            var self = this,
                $field = this.$field;

            // get repeater
            var $repeater = acf.get_closest_field(this.$field, 'repeater');

            // popup
            var frame = acf.media.popup({

                title: acf._e('image', 'select'),
                mode: 'select',
                type: 'image',
                field: acf.get_field_key($field),
                multiple: $repeater.exists(),
                library: this.o.library,
                mime_types: this.o.mime_types,

                select: function(attachment, i) {

                    // select / add another image field?
                    if (i > 0) {
                        // vars
                        var key = acf.get_field_key($field),
                            $tr = $field.closest('.acf-row');

                        // reset field
                        $field = false;

                        // find next image field
                        $tr.nextAll('.acf-row:visible').each(function() {

                            // get next $field
                            $field = acf.get_field(key, $(this));

                            // bail early if $next was not found
                            if (!$field) {
                                return;
                            }

                            // bail early if next file uploader has value
                            if ($field.find('.acf-image-uploader.has-value').exists()) {
                                $field = false;
                                return;
                            }

                            // end loop if $next is found
                            return false;

                        });

                        // add extra row if next is not found
                        if (!$field) {

                            $tr = acf.fields.repeater.doFocus($repeater).add();

                            // bail early if no $tr (maximum rows hit)
                            if (!$tr) {
                                return false;
                            }

                            // get next $field
                            $field = acf.get_field(key, $tr);

                        }

                    }

                    // focus
                    self.doFocus($field);

                    // render
                    self.render(self.prepare(attachment));

                }

            });

        },

        prepare: function(attachment) {
            // vars
            var image = {
                id: attachment.id,
                url: attachment.attributes.url
            };

            //console.log( 'Hep!' );

            leftVal = this.$el.find('[data-name="acf-focuspoint-left"]').val();
            topVal = this.$el.find('[data-name="acf-focuspoint-top"]').val();

            if( leftVal && topVal ){
                this.$el.find('.focal-point-picker').css({
                    top: topVal + '%',
                    left: leftVal + '%'
                });
            }
            else {
                this.$el.find('.focal-point-picker').css({
                    top: '50%',
                    left: '50%'
                });
            }
            this.$el.find('.focuspoint-image.no-image').removeClass('no-image').addClass('has-image');

            // check for preview size
            if (acf.isset(attachment.attributes, 'sizes', this.o.preview_size, 'url')) {
                image.url = attachment.attributes.sizes[this.o.preview_size].url;
            }

            // return
            return image;

        },

        render: function(image) {

            // set atts
            this.$el.find('[data-name="acf-focuspoint-img"]').attr('src', image.url);
            this.$el.find('[data-name="acf-focuspoint-img-id"]').val(image.id).trigger('change');

            // set div class
            this.$el.addClass('has-value');

        },

        edit: function() {
            // reference
            var self = this;

            // popup
            var frame = acf.media.popup({

                title: acf._e('image', 'edit'),
                type: 'image',
                button: acf._e('image', 'update'),
                mode: 'edit',
                id: id,

                select: function(attachment, i) {
                    self.render(self.prepare(attachment)).trigger('change');
                }

            });

        },
        
        removeAttachment: function(){
            this.render( false );
        },

        remove: function() {
            // vars
            var attachment = {
                id: '',
                url: ''
            };

            this.removeAttachment();

            // remove position
            this.$el.find('[data-name="acf-focuspoint-top"]').val('');
            this.$el.find('[data-name="acf-focuspoint-left"]').val('');

            // remove class
            this.$el.find('.focuspoint-image.has-image').removeClass('has-image').addClass('no-image');
            this.$el.removeClass('has-value').trigger('change');

        },

        change: function(e) {
            this.$el.find('[data-name="acf-focuspoint-img-id"]').val(e.$el.val()).trigger('change');
        }

    });


    if (typeof acf.add_action !== 'undefined') {

        /*
         *  ready append (ACF5)
         *
         *  These are 2 events which are fired during the page load
         *  ready = on page load similar to $(document).ready()
         *  append = on new DOM elements appended via repeater field
         *
         *  @type    event
         *  @date    20/07/13
         *
         *  @param   $el (jQuery selection) the jQuery element which contains the ACF fields
         *  @return  n/a
         */

        acf.add_action('ready append', function($el) {

            // search $el for fields of type 'focuspoint'
            acf.get_fields({ type: 'focuspoint' }, $el).each(function() {

                //initialize_field($(this));
                acf.fields.focuspoint.initialize($el)

            });

        });


    } else {


        /*
         *  acf/setup_fields (ACF4)
         *
         *  This event is triggered when ACF adds any new elements to the DOM.
         *
         *  @type    function
         *  @since   1.0.0
         *  @date    01/01/12
         *
         *  @param   event       e: an event object. This can be ignored
         *  @param   Element     postbox: An element which contains the new HTML
         *
         *  @return  n/a
         */

        $(document).live('acf/setup_fields', function(e, postbox) {

            $(postbox).find('.field[data-field_type="focuspoint"]').each(function() {

                initialize_field($(this));

            });

        });

    }

    // Initialize dynamic block preview (editor).
    if (window.acf) {
        //window.acf.addAction('render_block_preview/type=focuspoint', initialize_field);
        window.acf.addAction('render_block_preview/type=focuspoint', acf.fields.focuspoint.initialize);
    }


})(jQuery);

fields/class-acffp-acf-field-focuspoint-v5.php
<?php

// exit if accessed directly
if( ! defined( 'ABSPATH' ) ) exit;


// check if class already exists
if( !class_exists('acffp_acf_field_focuspoint') ) :


class acffp_acf_field_focuspoint extends acf_field {
	
	
	/*
	*  __construct
	*
	*  This function will setup the field type data
	*
	*  @type	function
	*  @date	5/03/2014
	*  @since	5.0.0
	*
	*  @param	n/a
	*  @return	n/a
	*/
	
	function __construct( $settings ) {
		
		/*
		*  name (string) Single word, no spaces. Underscores allowed
		*/
		
		$this->name = 'focuspoint';
		
		
		/*
		*  label (string) Multiple words, can include spaces, visible when selecting a field type
		*/
		
		$this->label = __('FocusPoint', 'acffp');
		
		
		/*
		*  category (string) basic | content | choice | relational | jquery | layout | CUSTOM GROUP NAME
		*/
		
		$this->category = 'jquery';
		
		
		/*
		*  defaults (array) Array of default settings which are merged into the field object. These are used later in settings
		*/
		
		$this->defaults = array(
			'preview_size'	=>	'large',
			'library'		=> 'all',
			'mime_types'	=> '',
		);
		
		
		/*
		*  l10n (array) Array of strings that are used in JavaScript. This allows JS strings to be translated in PHP and loaded via:
		*  var message = acf._e('focuspoint', 'error');
		*/
		
		$this->l10n = array();
		
		
		/*
		*  settings (array) Store plugin settings (url, path, version) as a reference for later use with assets
		*/
		
		$this->settings = $settings;
		

		// do not delete!
    	parent::__construct();


	}
	
	
	/*
	*  render_field_settings()
	*
	*  Create extra settings for your field. These are visible when editing a field
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$field (array) the $field being edited
	*  @return	n/a
	*/
	
	function render_field_settings( $field ) {
		
		/*
		*  acf_render_field_setting
		*
		*  This function will create a setting for your field. Simply pass the $field parameter and an array of field settings.
		*  The array of settings does not require a `value` or `prefix`; These settings are found from the $field array.
		*
		*  More than one setting can be added by copy/paste the above code.
		*  Please note that you must also have a matching $defaults value for the field name (font_size)
		*/
		
		// clear numeric settings
		
		// clear numeric settings
		$clear = array(
			'min_width',
			'min_height',
			'min_size',
			'max_width',
			'max_height',
			'max_size'
		);
		
		foreach( $clear as $k ) {
			if( empty($field[$k]) ) {
				$field[$k] = '';
			}
		}
		
		// Preview size select
		acf_render_field_setting( $field, array(
			'label'			=> __('Preview Size','acf-focuspoint'),
			'instructions'	=> __('Image used to create a FocusPoint. Should be around the same image ratio as Image Size','acf-focuspoint'),
			'type'			=> 'select',
			'name'			=> 'preview_size',
			'choices'		=>	acf_get_image_sizes()
		));
		
		
		// library
		acf_render_field_setting( $field, array(
			'label'			=> __('Library','acf'),
			'instructions'	=> __('Limit the media library choice','acf'),
			'type'			=> 'radio',
			'name'			=> 'library',
			'layout'		=> 'horizontal',
			'choices' 		=> array(
				'all'			=> __('All', 'acf'),
				'uploadedTo'	=> __('Uploaded to post', 'acf')
			)
		));
		
		
		// Min sizes
		acf_render_field_setting( $field, array(
			'label'			=> __('Minimum', 'acf-focuspoint'),
			'instructions'	=> __('Restrict which images can be uploaded. Note: this is not working yet!', 'acf-focuspoint'),
			'type'			=> 'text',
			'name'			=> 'min_width',
			'prepend'		=> __('Width', 'acf-focuspoint'),
			'append'		=> 'px',
		));
		
		acf_render_field_setting( $field, array(
			'label'			=> '',
			'type'			=> 'text',
			'name'			=> 'min_height',
			'prepend'		=> __('Height', 'acf-focuspoint'),
			'append'		=> 'px',
			'_append' 		=> 'min_width'
		));
		
		acf_render_field_setting( $field, array(
			'label'			=> '',
			'type'			=> 'text',
			'name'			=> 'min_size',
			'prepend'		=> __('File size', 'acf-focuspoint'),
			'append'		=> 'MB',
			'_append' 		=> 'min_width'
		));	
		
		
		// Max sizes
		acf_render_field_setting( $field, array(
			'label'			=> __('Maximum', 'acf-focuspoint'),
			'instructions'	=> __('Restrict which images can be uploaded. Note: this is not working yet!', 'acf-focuspoint'),
			'type'			=> 'text',
			'name'			=> 'max_width',
			'prepend'		=> __('Width', 'acf-focuspoint'),
			'append'		=> 'px',
		));
		
		acf_render_field_setting( $field, array(
			'label'			=> '',
			'type'			=> 'text',
			'name'			=> 'max_height',
			'prepend'		=> __('Height', 'acf-focuspoint'),
			'append'		=> 'px',
			'_append' 		=> 'max_width'
		));
		
		acf_render_field_setting( $field, array(
			'label'			=> '',
			'type'			=> 'text',
			'name'			=> 'max_size',
			'prepend'		=> __('File size', 'acf-focuspoint'),
			'append'		=> 'MB',
			'_append' 		=> 'max_width'
		));	
		
		
		// allowed type
		acf_render_field_setting( $field, array(
			'label'			=> __('Allowed file types','acf'),
			'instructions'	=> __('Comma separated list. Leave blank for all types','acf'),
			'type'			=> 'text',
			'name'			=> 'mime_types',
		));

	}
	
	
	
	/*
	*  render_field()
	*
	*  Create the HTML interface for your field
	*
	*  @param	$field (array) the $field being rendered
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$field (array) the $field being edited
	*  @return	n/a
	*/
	
	function render_field( $field ) {

		// Merge defaults
		$field = array_merge($this->defaults, $field);
		
		// Get set image id
		$id = (isset($field['value']['id'])) ? $field['value']['id'] : '';

		// data vars
		$data = array(
			'top'		=>	isset($field['value']['top']) ? $field['value']['top'] : '',
			'left'		=>	isset($field['value']['left']) ? $field['value']['left'] : '',
		);
		
		// If we already have an image set...
		if ($id) {
			
			// Get image by ID, in size set via options
			$img = wp_get_attachment_image_src($id, $field['preview_size']);
						
		}
			
		// If image found...
		// Set to hide add image button / show canvas
		$is_active 	= ($id) ? 'active' : '';

		// And set src
		$url = ($id) ? $img[0] : '';
		
		// create Field HTML
		?>

		<div class="acf-focuspoint acf-image-uploader <?php echo $is_active; ?>" data-preview_size="<?php echo $field['preview_size']; ?>" data-library="<?php echo $field['library']; ?>" data-mime_types="<?php echo $field['mime_types']; ?>">

			<input data-name="acf-focuspoint-img-id" type="hidden" name="<?php echo $field['name']; ?>[id]" value="<?php echo $id; ?>" />

			<?php foreach ($data as $k => $d): ?>
				<input data-name="acf-focuspoint-<?php echo $k ?>" type="hidden" name="<?php echo $field['name']; ?>[<?php echo $k ?>]" value="<?php echo $d ?>" />
			<?php endforeach ?>

			<div class="focuspoint-image <?php echo $id && wp_attachment_is_image( $id ) ? 'has-image' : 'no-image' ?>">
				<img data-name="acf-focuspoint-img" src="<?php echo $url; ?>">
				<img class="focal-point-picker" src="<?php echo $this->settings['url']; ?>assets/images/focal-point-picker.svg" style="top: <?php echo $data['top']; ?>%; left: <?php echo $data['left']; ?>%;">
				<div class="focuspoint-selection-layer"></div>
				<a class="acf-button-delete acf-icon -cancel acf-icon-cancel dark" data-name="remove"></a>
			</div>
			
			<div class="view hide-if-value">
			    <p><?php _e('No image selected','acf'); ?> <a data-name="add" class="acf-button button" href="#"><?php _e('Add Image','acf'); ?></a></p>
			</div>

		</div>

		<?php

	}
	
		
	/*
	*  input_admin_enqueue_scripts()
	*
	*  This action is called in the admin_enqueue_scripts action on the edit screen where your field is created.
	*  Use this action to add CSS + JavaScript to assist your render_field() action.
	*
	*  @type	action (admin_enqueue_scripts)
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	n/a
	*  @return	n/a
	*/
	function input_admin_enqueue_scripts() {
		
		// vars
		$url = $this->settings['url'];
		$version = $this->settings['version'];
		
		
		// register & include JS
		wp_register_script('acffp', "{$url}assets/js/input.min.js", array('jquery', 'acf-input'), $version );
		wp_enqueue_script('acffp');
    		wp_enqueue_media();
		
		
		// register & include CSS
		wp_register_style('acffp', "{$url}assets/css/input.min.css", array('acf-input'), $version );
		wp_enqueue_style('acffp');
		
	}
	
	
	
	
	/*
	*  input_admin_head()
	*
	*  This action is called in the admin_head action on the edit screen where your field is created.
	*  Use this action to add CSS and JavaScript to assist your render_field() action.
	*
	*  @type	action (admin_head)
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	n/a
	*  @return	n/a
	*/

	/*
		
	function input_admin_head() {
	
		
		
	}
	
	*/
	
	
	/*
   	*  input_form_data()
   	*
   	*  This function is called once on the 'input' page between the head and footer
   	*  There are 2 situations where ACF did not load during the 'acf/input_admin_enqueue_scripts' and 
   	*  'acf/input_admin_head' actions because ACF did not know it was going to be used. These situations are
   	*  seen on comments / user edit forms on the front end. This function will always be called, and includes
   	*  $args that related to the current screen such as $args['post_id']
   	*
   	*  @type	function
   	*  @date	6/03/2014
   	*  @since	5.0.0
   	*
   	*  @param	$args (array)
   	*  @return	n/a
   	*/
   	
   	/*
   	
   	function input_form_data( $args ) {
	   	
		
	
   	}
   	
   	*/
	
	
	/*
	*  input_admin_footer()
	*
	*  This action is called in the admin_footer action on the edit screen where your field is created.
	*  Use this action to add CSS and JavaScript to assist your render_field() action.
	*
	*  @type	action (admin_footer)
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	n/a
	*  @return	n/a
	*/

	/*
		
	function input_admin_footer() {
	
		
		
	}
	
	*/
	
	
	/*
	*  field_group_admin_enqueue_scripts()
	*
	*  This action is called in the admin_enqueue_scripts action on the edit screen where your field is edited.
	*  Use this action to add CSS + JavaScript to assist your render_field_options() action.
	*
	*  @type	action (admin_enqueue_scripts)
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	n/a
	*  @return	n/a
	*/

	/*
	
	function field_group_admin_enqueue_scripts() {
		
	}
	
	*/

	
	/*
	*  field_group_admin_head()
	*
	*  This action is called in the admin_head action on the edit screen where your field is edited.
	*  Use this action to add CSS and JavaScript to assist your render_field_options() action.
	*
	*  @type	action (admin_head)
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	n/a
	*  @return	n/a
	*/

	/*
	
	function field_group_admin_head() {
	
	}
	
	*/


	/*
	*  load_value()
	*
	*  This filter is applied to the $value after it is loaded from the db
	*
	*  @type	filter
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$value (mixed) the value found in the database
	*  @param	$post_id (mixed) the $post_id from which the value was loaded
	*  @param	$field (array) the field array holding all the field options
	*  @return	$value
	*/
	
	/*
	
	function load_value( $value, $post_id, $field ) {
		
		return $value;
		
	}
	
	*/
	
	
	/*
	*  update_value()
	*
	*  This filter is applied to the $value before it is saved in the db
	*
	*  @type	filter
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$value (mixed) the value found in the database
	*  @param	$post_id (mixed) the $post_id from which the value was loaded
	*  @param	$field (array) the field array holding all the field options
	*  @return	$value
	*/
	
	function update_value( $value, $post_id, $field ) {
		
		if( empty( $value['id'] ) ){
			return false;
		}
		if( empty( $value['left'] ) && empty( $value['top'] ) ){
			$value['left'] = 50;
			$value['top'] = 50;
		}

		return $value;
		
	}
	
	
	
	
	/*
	*  format_value()
	*
	*  This filter is appied to the $value after it is loaded from the db and before it is returned to the template
	*
	*  @type	filter
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$value (mixed) the value which was loaded from the database
	*  @param	$post_id (mixed) the $post_id from which the value was loaded
	*  @param	$field (array) the field array holding all the field options
	*
	*  @return	$value (mixed) the modified value
	*/
		
	/*
	
	function format_value( $value, $post_id, $field ) {
		
		// bail early if no value
		if( empty($value) ) {
		
			return $value;
			
		}
		
		
		// apply setting
		if( $field['font_size'] > 12 ) { 
			
			// format the value
			// $value = 'something';
		
		}
		
		
		// return
		return $value;
	}
	
	*/
	
	
	
	/*
	*  validate_value()
	*
	*  This filter is used to perform validation on the value prior to saving.
	*  All values are validated regardless of the field's required setting. This allows you to validate and return
	*  messages to the user if the value is not correct
	*
	*  @type	filter
	*  @date	11/02/2014
	*  @since	5.0.0
	*
	*  @param	$valid (boolean) validation status based on the value and the field's required setting
	*  @param	$value (mixed) the $_POST value
	*  @param	$field (array) the field array holding all the field options
	*  @param	$input (string) the corresponding input name for $_POST value
	*  @return	$valid
	*/
	
	
	
	function validate_value( $valid, $value, $field, $input ){

		// vd( $valid );
		// vdd( empty($value['id']) );

		// Bail early if field not required and value not set
		if( $field['required'] === 0 && !$value['id'] ) return true;
		
		// bail early if field required and id empty		
		if( $field['required'] !== 0 && empty($value['id']) ) return false;
		
		// bail ealry if id not numeric
		if( !is_numeric($value['id']) ) return false;

		$image = wp_get_attachment_image_src( $value['id'], 'full' );
		$image_size_kb = filesize( get_attached_file( $value['id'] ) );
		
		if( !empty($field['min_width']) && $image[1] < $field['min_width'] ) {
			$valid = sprintf( __('Image width must be at least %dpx.', 'acf-focuspoint'), $field['min_width'], $image[1], $image[2] );
		}
		elseif( !empty($field['min_height']) && $image[2] < $field['min_height'] ) {
			$valid = sprintf( __('Image height must be at least %dpx.', 'acf-focuspoint'), $field['min_height'], $image[1], $image[2] );
		}
		elseif( !empty($field['min_size']) && $image_size_kb < $field['min_size'] * (1024 * 1024) ) {
			$valid = sprintf( __('File size must be at least %d&nbsp;KB.', 'acf-focuspoint'), size_format( $field['min_size'] * (1024 * 1024), 2), size_format( $image_size_kb, 2 ) );
		}
		elseif( !empty($field['max_width']) && $image[1] > $field['max_width'] ) {
			$valid = sprintf( __('Image width must not exceed %dpx.', 'acf-focuspoint'), $field['max_width'], $image[1], $image[2] );
		}
		elseif( !empty($field['max_height']) && $image[2] > $field['max_height'] ) {
			$valid = sprintf( __('Image height must not exceed %dpx.', 'acf-focuspoint'), $field['max_height'], $image[1], $image[2] );
		}
		elseif( !empty($field['max_size']) && $image_size_kb > $field['max_size'] * (1024 * 1024) ) {
			$valid = sprintf( __('File size must not exceed %d&nbsp;KB.', 'acf-focuspoint'), size_format( $field['max_size'] * (1024 * 1024), 2), size_format( $image_size_kb, 2 ) );
		}
		else{
			$valid = true;
		}
		
		// return
		return $valid;

	}
	
	
	
	
	/*
	*  delete_value()
	*
	*  This action is fired after a value has been deleted from the db.
	*  Please note that saving a blank value is treated as an update, not a delete
	*
	*  @type	action
	*  @date	6/03/2014
	*  @since	5.0.0
	*
	*  @param	$post_id (mixed) the $post_id from which the value was deleted
	*  @param	$key (string) the $meta_key which the value was deleted
	*  @return	n/a
	*/
	
	/*
	
	function delete_value( $post_id, $key ) {
		
		
		
	}
	
	*/
	
	
	/*
	*  load_field()
	*
	*  This filter is applied to the $field after it is loaded from the database
	*
	*  @type	filter
	*  @date	23/01/2013
	*  @since	3.6.0	
	*
	*  @param	$field (array) the field array holding all the field options
	*  @return	$field
	*/
	
	/*
	
	function load_field( $field ) {
		
		return $field;
		
	}	
	
	*/
	
	
	/*
	*  update_field()
	*
	*  This filter is applied to the $field before it is saved to the database
	*
	*  @type	filter
	*  @date	23/01/2013
	*  @since	3.6.0
	*
	*  @param	$field (array) the field array holding all the field options
	*  @return	$field
	*/
	
	/*
	
	function update_field( $field ) {
		
		return $field;
		
	}	
	
	*/
	
	
	/*
	*  delete_field()
	*
	*  This action is fired after a field is deleted from the database
	*
	*  @type	action
	*  @date	11/02/2014
	*  @since	5.0.0
	*
	*  @param	$field (array) the field array holding all the field options
	*  @return	n/a
	*/
	
	/*
	
	function delete_field( $field ) {
		
		
		
	}	
	
	*/
	
	
}


// initialize
new acffp_acf_field_focuspoint( $this->settings );


// class_exists check
endif;

?>
